### PR TITLE
Case 21449: [AvatarPackager/AvatarDoctor] Warnings update

### PR DESCRIPTION
**Pull request by thoys**
Adresses the following PRs:
- https://highfidelity.manuscript.com/f/cases/21449/Avatar-doc-improvement-Warnings-update
- https://highfidelity.manuscript.com/f/cases/21284/AvatarPackager-AvatarDoctor-Scaling-of-bones-should-be-applied-properly

Depends on PRs:
- ~https://github.com/highfidelity/hifi/pull/15049~ (merged)
- ~https://github.com/highfidelity/hifi/pull/15013~ (merged)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17647/HighFidelity-Beta-PR15035-fea8b1b.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17647/HighFidelity-Quest-Beta-PR15035-fea8b1b.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17647/HighFidelity-Beta-PR15035-fea8b1b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17647/HighFidelity-Beta-Interface-PR15035-fea8b1b.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17647/HighFidelity-Beta-PR15035-fea8b1b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17647/HighFidelity-Beta-Interface-PR15035-fea8b1b.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17763/HighFidelity-Beta-PR15035-9678f7a.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17763/HighFidelity-Quest-Beta-PR15035-9678f7a.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17763/HighFidelity-Beta-PR15035-9678f7a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17763/HighFidelity-Beta-Interface-PR15035-9678f7a.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17906/HighFidelity-Beta-PR15035-95cfc4c.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17906/HighFidelity-Quest-Beta-PR15035-95cfc4c.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17906/HighFidelity-Beta-PR15035-95cfc4c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17906/HighFidelity-Beta-Interface-PR15035-95cfc4c.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17909/HighFidelity-Beta-PR15035-3db0556.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17909/HighFidelity-Quest-Beta-PR15035-3db0556.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17909/HighFidelity-Beta-PR15035-3db0556.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17909/HighFidelity-Beta-Interface-PR15035-3db0556.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17909/HighFidelity-Beta-PR15035-3db0556.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17909/HighFidelity-Beta-Interface-PR15035-3db0556.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17937/HighFidelity-Beta-PR15035-1b6c6c9.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17937/HighFidelity-Beta-Interface-PR15035-1b6c6c9.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17937/HighFidelity-Beta-PR15035-1b6c6c9.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17937/HighFidelity-Quest-Beta-PR15035-1b6c6c9.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17937/HighFidelity-Beta-PR15035-1b6c6c9.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17937/HighFidelity-Beta-Interface-PR15035-1b6c6c9.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17980/HighFidelity-Beta-PR15035-f2bca5d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17980/HighFidelity-Beta-Interface-PR15035-f2bca5d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17980/HighFidelity-Beta-PR15035-f2bca5d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17980/HighFidelity-Beta-Interface-PR15035-f2bca5d.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17980/HighFidelity-Beta-PR15035-f2bca5d.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17980/HighFidelity-Quest-Beta-PR15035-f2bca5d.apk)

File: interface/src/avatar/AvatarDoctor.cpp [hunk 1]
```diff
@@ -39,136 +81,216 @@ void AvatarDoctor::startDiagnosing() {
...
-            auto modelTexturesURLs = model->getTextures();
-            _externalTextureCount = externalTextures.length();
-            foreach(const QString textureKey, externalTextures) {
-                if (!modelTexturesURLs.contains(textureKey)) {
-                    _missingTextureCount++;
-                    _checkedTextureCount++;
-                    continue;
+            // Multiple skeleton root joints checkup
+            int skeletonRootJoints = 0;
+            foreach(const HFMJoint& joint, avatarModel.joints) {
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Style: space after `foreach`

File: interface/src/avatar/AvatarDoctor.cpp [hunk 2]
```diff
@@ -39,136 +81,216 @@ void AvatarDoctor::startDiagnosing() {
...
 
-                const QUrl textureURL = modelTexturesURLs[textureKey].toUrl();
+            if (skeletonRootJoints > 1) {
+                _errors.push_back({ "Multiple root joints found.", DEFAULT_URL });
+            }
 
-                auto textureResource = DependencyManager::get<TextureCache>()->getTexture(textureURL);
-                auto checkTextureLoadingComplete = [this, DEFAULT_URL] () mutable {
-                    qDebug() << "checkTextureLoadingComplete" << _checkedTextureCount << "/" << _externalTextureCount;
+            const auto rig = new Rig();
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Should we create this on the stack?

File: interface/src/avatar/AvatarDoctor.cpp [hunk 3]
```diff
@@ -39,136 +81,216 @@ void AvatarDoctor::startDiagnosing() {
...
+                }
+            }
+            rig->deleteLater();
+
+            auto mapping = resource->getMapping();
+
+            if (mapping.contains(JOINT_NAME_MAPPING_FIELD) && mapping[JOINT_NAME_MAPPING_FIELD].type() == QVariant::Hash) {
+                const auto& jointNameMappings = mapping[JOINT_NAME_MAPPING_FIELD].toHash();
+                QStringList jointValues;
+                foreach(const auto& jointVariant, jointNameMappings.values()) {
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Style: space after `foreach`

File: interface/src/avatar/AvatarDoctor.cpp [hunk 4]
```diff
@@ -39,136 +81,216 @@ void AvatarDoctor::startDiagnosing() {
...
+            // Continue diagnosing the textures as soon as the material mappings have tried to load.
+            if (_materialMappingLoadedCount == _materialMappingCount) {
+                // TEXTURES
+                diagnoseTextures();
+            }
+        };
+
+        _materialMappingCount = (int)model->getMaterialMapping().size();
+        _materialMappingLoadedCount = 0;
+        foreach(const auto& materialMapping, model->getMaterialMapping()) {
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Style: space after `foreach`

File: interface/src/avatar/AvatarDoctor.cpp [hunk 5]
```diff
@@ -184,6 +306,115 @@ void AvatarDoctor::startDiagnosing() {
...
+        if (texture.filename.isEmpty()) {
+            return;
+        }
+        if (texture.content.isEmpty() && !externalTextures.contains(texture.name)) {
+            externalTextures << texture.name;
+        }
+        texturesFound++;
+    };
+
+    foreach(const HFMMaterial material, avatarModel.materials) {
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Style: space after `foreach`

File: interface/src/avatar/AvatarDoctor.cpp [hunk 6]
```diff
@@ -184,6 +306,115 @@ void AvatarDoctor::startDiagnosing() {
...
+        addTextureToList(material.roughnessTexture);
+        addTextureToList(material.specularTexture);
+        addTextureToList(material.metallicTexture);
+        addTextureToList(material.emissiveTexture);
+        addTextureToList(material.occlusionTexture);
+        addTextureToList(material.scatteringTexture);
+        addTextureToList(material.lightmapTexture);
+    }
+
+    foreach(const auto& materialMapping, model->getMaterialMapping()) {
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Style: space after `foreach`

File: interface/src/avatar/AvatarDoctor.cpp [hunk 7]
```diff
@@ -184,6 +306,115 @@ void AvatarDoctor::startDiagnosing() {
...
+        foreach(const auto& networkMaterial, materialMapping.second.data()->parsedMaterials.networkMaterials) {
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Style: space after `foreach`

File: interface/src/avatar/AvatarDoctor.cpp [hunk 8]
```diff
@@ -184,6 +306,115 @@ void AvatarDoctor::startDiagnosing() {
     }    
 }
 
+void AvatarDoctor::diagnoseTextures() {
+    const auto model = _model.data();
+    const auto avatarModel = _model.data()->getHFMModel();
+    QStringList externalTextures{};
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: I would generally suggest reaching for `QVector` or `std::vector` first. From a performance perspective, unless you have really specific needs, a vector is going to perform better.

File: interface/src/avatar/AvatarDoctor.cpp [hunk 9]
```diff
@@ -184,6 +306,115 @@ void AvatarDoctor::startDiagnosing() {
...
+
+    if (texturesFound == 0) {
+        _errors.push_back({ tr("No textures assigned."), DEFAULT_URL });
+    }
+
+    if (!externalTextures.empty()) {
+        // Check External Textures:
+        auto modelTexturesURLs = model->getTextures();
+        _externalTextureCount = externalTextures.length();
+        foreach(const QString textureKey, externalTextures) {
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Style: space after `foreach`

File: interface/src/avatar/AvatarDoctor.cpp [hunk 10]
```diff
@@ -184,6 +306,115 @@ void AvatarDoctor::startDiagnosing() {
...
+                _missingTextureCount++;
+                _checkedTextureCount++;
+                continue;
+            }
+
+            const QUrl textureURL = modelTexturesURLs[textureKey].toUrl();
+
+            auto textureResource = DependencyManager::get<TextureCache>()->getTexture(textureURL);
+            auto checkTextureLoadingComplete = [this]() mutable {
+                qDebug() << "checkTextureLoadingComplete" << _checkedTextureCount << "/" << _externalTextureCount;
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: delete this?

File: interface/src/avatar/AvatarDoctor.cpp [hunk 11]
```diff
@@ -184,6 +306,115 @@ void AvatarDoctor::startDiagnosing() {
...
+                textureResource->refresh();
+                if (textureResource->isLoaded()) {
+                    textureLoaded(!textureResource->isFailed());
+                } else {
+                    connect(textureResource.data(), &NetworkTexture::finished, this, textureLoaded);
+                }
+            } else {
+                _missingTextureCount++;
+                _checkedTextureCount++;
+                checkTextureLoadingComplete();
```
Comment (huffman) [inline interface/src/avatar/AvatarDoctor.cpp]: Should we just check this once, after we've iterated through all of the textures?

Comment (thoys) [reply interface/src/avatar/AvatarDoctor.cpp]: But how would we know if we've had them all? I made the code more DRY by checking it in a  lamda function :)

Comment (thoys) [reply interface/src/avatar/AvatarDoctor.cpp]: Oh I see what you mean, just for this case here. could do, will have to make the code more failsafe then though.

Comment (huffman) [reply interface/src/avatar/AvatarDoctor.cpp]: It's not critical. I wouldn't make large changes because of this.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18013/HighFidelity-Beta-PR15035-06e5927.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18013/HighFidelity-Beta-Interface-PR15035-06e5927.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18013/HighFidelity-Beta-PR15035-06e5927.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18013/HighFidelity-Quest-Beta-PR15035-06e5927.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18013/HighFidelity-Beta-PR15035-06e5927.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18013/HighFidelity-Beta-Interface-PR15035-06e5927.exe

**Pull request merged**